bundles/framework/layerlist/view/LayerViewTabs/LayerList/LayerListHandler.js

```diff
@@ -15,94 +15,114 @@ class UIStateHandler extends StateHandler {
 
-        this._initFilterHandler();
-        this._initLayerCollapseHandlers(groupingOptions);
+        this.filterHandler = this._createFilterHandler();
+        this.layerCollapseHandlers = this._createLayerCollapseHandlers(groupingOptions);
+        const selectedGrouping = groupingOptions[0].getKey();
+        const collapseHandler = this._getCollapseHandler(selectedGrouping);
 
         this.state = {
             loading: false,
+            updating: false,
+            showAddButton: this.sandbox.hasHandler('ShowLayerEditorRequest'),
             grouping: {
-                selected: groupingOptions[0].getKey(),
+                selected: selectedGrouping,
                 options: groupingOptions
+            },
+            filter: {
+                state: this.filterHandler.getState(),
+                mutator: this.filterHandler.getMutator()
+            },
+            collapse: {
+                state: collapseHandler.getState(),
+                mutator: collapseHandler.getMutator()
             }
         };
-        this._updateFilter();
-        this._updateCollapse();
     }
 
-    _initLayerCollapseHandlers (groupingOptions) {
+    _createLayerCollapseHandlers (groupingOptions) {
         const handlers = {};
         groupingOptions.forEach(option => {
             const handler = new LayerCollapseHandler(this.instance, option.getMethod());
-            handler.addStateListener(() => {
+            handler.addStateListener(state => {
                 if (option.getKey() !== this.state.grouping.selected) {
                     // Not the active grouping, ignore.
                     return;
                 }
-                this._updateCollapse();
+                this.updateState({
+                    collapse: {
+                        state,
+                        mutator: handler.getMutator()
+                    },
+                    updating: false
+                });
             });
             handlers[option.getKey()] = handler;
         });
-        this.layerCollapseHandlers = handlers;
+        return handlers;
     }
 
-    _initFilterHandler () {
-        const uiHandler = new FilterHandler(this.instance);
+    _createFilterHandler () {
+        const handler = new FilterHandler(this.instance);
 
-        let previousFilter = null;
-        const updateLayerFilters = (filterState) => {
-            previousFilter = filterState;
-            const { activeFilterId, searchText } = filterState;
-            setTimeout(() => {
-                this._getCollapseHandler().setFilter(activeFilterId, searchText);
-            }, UI_UPDATE_TIMEOUT);
+        let previousState = null;
+        const updateLayerFilters = () => {
+            previousState = handler.getState();
+            const { activeFilterId, searchText } = previousState;
+            this.updateState({ updating: true });
+            setTimeout(() => this._getCollapseHandler().setFilter(activeFilterId, searchText), UI_UPDATE_TIMEOUT);
```

Comment:
This might trigger multiple timeouts if the user uses arrow keys for browsing through the filters? Not sure if it's an issue though